### PR TITLE
Maxim.task 4

### PR DESCRIPTION
**Pull request by PolyMaG**
ProductSpider

### CODE REVIEW

File: mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    def parse(self, response):
+        """
+        Find links to product page.
+        """
+        product_page_links = response.css('.detailsLink')
+        yield from response.follow_all(product_page_links, self.parse_product)
+
+        pagination_links = response.css('span.fleft a')
+        yield from response.follow_all(pagination_links, self.parse)
```
Comment (DmitryPoliuha) [inline mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py:30]: nice :+1:

File: mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+from olx_scraping.items import Product
+
+
+class ProductSpider(scrapy.Spider):
+    name = 'products'
+
+    custom_settings = {
+        'FEED_URI': 'output.json',
+        'FEED_EXPORT_ENCODING': 'utf-8',
+        'CLOSESPIDER_ITEMCOUNT': 10,
```
Comment (DmitryPoliuha) [inline mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py:13]: let's leave only one setting: settings.py or custom_settings here

File: mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        pagination_links = response.css('span.fleft a')
+        yield from response.follow_all(pagination_links, self.parse)
+
+    def parse_product(self, response):
+        """
+        Extract data from product page
+        """
+        def extract_with_css(query):
+            return response.css(query).get(default='').strip()
```
Comment (DmitryPoliuha) [inline mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py:37]: :+1:

File: mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+        if self.count < self.MAX_COUNT:
+            product = Product()
+            product['name'] = extract_with_css('h1::text')
+            product['url'] = response.xpath('/html/head/link[1]/@href').get()
```
Comment (DmitryPoliuha) [inline mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py:42]: I'm sure that it is possible to extract URL from response object without using css or xpath

File: mpolyakov/homework_4/olx_scraping/output.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+{"name": "IQOS 3 duos", "url": "https://www.olx.ua/obyavlenie/iqos-3-duos-IDJAYhX.html", "price": 599, "currency": " грн.", "category": "Б/у", "state": "Б/у", "description": "Айкос 3 дуос", "date_of_creation": "в 09:14, 23 октября 2020", "views_count": "9", "id": "673746681", "photos_urls": "https://ireland.apollo.olxcdn.com:443/v1/files/3yzpcgn56j9h1-UA/image;s=1000x700", "seller_name": "Анатолий", "seller_address": "Киев, Киевская область, Деснянский", "olx_delivery_availability": ""}
```
Comment (DmitryPoliuha) [inline mpolyakov/homework_4/olx_scraping/output.json:9]: category and state are the same? 🤔

Comment (DmitryPoliuha) [reply mpolyakov/homework_4/olx_scraping/output.json:9]: I'm sure that photos_urls should contain array of urls

Comment (DmitryPoliuha) [reply mpolyakov/homework_4/olx_scraping/output.json:9]: views_count should be int, I guess

Comment (DmitryPoliuha) [reply mpolyakov/homework_4/olx_scraping/output.json:9]: what will happen with the price in case of USD price?

**Review by DmitryPoliuha — COMMENTED** [review]
Looks good. `.pyc` files should be deleted. Some issues with parsing should be updated

File: mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py (hunk 1 above)
Comment (DmitryPoliuha) [reply mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py:30]: btw, do you know the meaning of the yield from?

File: mpolyakov/homework_4/olx_scraping/output.json (hunk 5 above)
Comment (PolyMaG) [reply mpolyakov/homework_4/olx_scraping/output.json:9]: It works: "currency": " $",

File: mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py (hunk 1 above)
Comment (PolyMaG) [reply mpolyakov/homework_4/olx_scraping/olx_scraping/spiders/product_spider.py:24]: It's the same as:
`for a in response.css('span.fleft a'):
    yield response.follow(a, self.parse_product)`

**Review by VChernyshoff — APPROVED** [review]

**Review by DmitryPoliuha — APPROVED** [review]
Nice

**Pull request merged**